README.md

```diff
@@ -10,12 +10,93 @@ Basic interface for manifolds in Julia.
+should keep in mind that also `exp!(M,p,p,X)` should be possible without side effects.
+
+The interface provides a way to determine the allocation type and a result (see [`allocate_result`]()) to compute/allocate
+the resulting memory, such that the default implementation allocating functions, like `exp` is to allocate the resulting memory and call `exp!`.
+
+!!! note
+   it might be useful to provide two distinct implementations, for example when using AD schemes.
+   The default is meant for ease of use (concerning implementation), since then one has to implement the mutating variants.
+
+### The higher level interface and ease of use
+
+The higher level interface should aim for a convenience layer that resolves defaults and
+creates fallbacks for certain input parameters, that have these properties.
+It usually should not dispatch on a certain manifold nor on certain point or (co- or tangent) vector types.
+
+This layer should also not resolve/dispatch from the allocating to the mutating variant.
+
+This is maybe best illustrated by two examples
+
+1. the exponential map usually has a long form, where one can specify a fraction (of `X`) where the evaluation should be. This is generically impülemented a
+
+  ```julia
+    exp(M::AbstractManifold, p, X, t::Real) = exp(M, p, t * X)
+  ```
+
+  On this level neither the manifold _nor_ the points should be too strictly typed, points and vectors should – for best of cases – never be types.
+
+2. for the retraction, a default retraction (usually exp) is specified/defined via [`default_retraction_method`]().
+  This also means, that the last parameter of
+
+  ```julia
+  retract(M::AbstractManifold, p, X, ::AbstractRetractionMethod=default_retraction_method(M))
+  ```
+
+  is optional and for a concrete type the dispatch on a certain retraction is done next.
+  To avoid ambiguities, this concrete type should always be the first argument we dispatch on:
+  The `ExponentialRetractionMethod` calls `exp`, any other retraction calls a function of different name without this last parameter,
+  for example the `PolarRetractionMethod` by default calls `retract_polar(M,p,X)`, which is actualy a function from the lower level, see next section.
+
+### The lower level interface – performance
+
+This lower level aims for performance, that is, any function should have as less as possible optional and keyword arguments
+and be typed as concrete as possible/necessary. This means
+
+* the function name should be similar to its high level parent (for example `retract` and `retract_polar`from above)
+* The manifold should always be a concrete manifold
+* the points/vectors should either be untyped (for the default representation of if there is only one) or provide all types concretely.
+
+A first thing to do on this level is the aforementioned default to pass from allocating to mutating functions.
+
+(TODO/Discuss - dispatch for decorators here instead of with the current/modular decorator pattern).
```

Comment:
Shouldn't dispatch over decorators be a part of the higher-level interface?

Reply:
Yes, or a middle layer in between, that is exactly where I am not yet sure - i.e. whether to first resolve the Retraction-default or first dispatch on the decorator.

But sure, in this area and (instead of the abstract types) Traits.
For the concrete ones the same as before (to distinguish different embeddins, more than the default with trait).

Reply:
Forwarding through decorators first seems to me like a good idea. Going from i.e. `retract` to `retract_polar` is moving into the lower level so I think transparent decorators should already be unwrapped at that point. Allocating-to-mutating default behavior should be more down the line than decorators because some manifolds need to override the allocation pattern (for example `Circle`).

Anyway, bases are much more complex than retractions and should perhaps get the most attention here.

Reply:
Sure, that is why I started with retractions, to first get used to the new ideas - will even do vector transports before I go over to bases.